File: sdk/typescript/examples/langgraphWrapper.ts
/**
 * LangGraph-style wrapper for AegisRun
 * 
 * This provides a graph-based execution model similar to LangGraph,
 * with full policy enforcement via AegisRun.
 */

import { AegisRunClient } from '../src/client';
import { Run } from '../src/run';
import { Step } from '../src/step';

interface State {
  messages: Array<{ role: string; content: string }>;
  currentStep: string;
  data: Record<string, any>;
}

type NodeFunction = (state: State) => Promise<State>;

class AegisGraph {
  private run: Run;
  private nodes: Map<string, NodeFunction> = new Map();
  private edges: Map<string, string> = new Map();
  private conditionalEdges: Map<string, (state: State) => string> = new Map();

  constructor(run: Run) {
    this.run = run;
  }

  addNode(name: string, fn: NodeFunction): this {
    this.nodes.set(name, fn);
    return this;
  }

  addEdge(fromNode: string, toNode: string): this {
    this.edges.set(fromNode, toNode);
    return this;
  }

  addConditionalEdge(
    fromNode: string,
    conditionFn: (state: State) => string
  ): this {
    this.conditionalEdges.set(fromNode, conditionFn);
    return this;
  }

  compile(): CompiledGraph {
    return new CompiledGraph(
      this.run,
      this.nodes,
      this.edges,
      this.conditionalEdges
    );
  }
}

class CompiledGraph {
  constructor(
    private run: Run,
    private nodes: Map<string, NodeFunction>,
    private edges: Map<string, string>,
    private conditionalEdges: Map<string, (state: State) => string>
  ) {}

  async invoke(initialState: State): Promise<State> {
    let state = initialState;
    let currentNode: string | undefined = 'start';
    const visitedNodes: string[] = [];

    while (currentNode && this.nodes.has(currentNode)) {
      const nodeFn = this.nodes.get(currentNode)!;

      // Execute node as a step with AegisRun policy enforcement
      state = await this.run.step(
        currentNode,
        { node: currentNode, state, visited: visitedNodes },
        async (step: Step) => {
          return await nodeFn(state);
        }
      );

      visitedNodes.push(currentNode);

      // Check for conditional edge first
      if (this.conditionalEdges.has(currentNode)) {
        const conditionFn = this.conditionalEdges.get(currentNode)!;
        currentNode = conditionFn(state);
      } else {
        // Otherwise use regular edge
        currentNode = this.edges.get(currentNode);
      }

      // Break if we've reached END or an unknown node
      if (currentNode === 'END' || currentNode === undefined) {
        break;
      }
    }

    return state;
  }
}

// Example usage demonstrating a ReAct-style agent
async function exampleReActAgent() {
  const client = new AegisRunClient({ baseUrl: 'http://localhost:8080' });

  const run = await new Run({
    client,
    agentId: 'react-agent',
    policyRef: 'react-agent-policy:v1',
    metadata: {
      agent_type: 'react',
      task: 'information_gathering',
    },
  }).start();

  console.log(`Run started: ${run.runId}`);

  // Define graph nodes
  const think = async (state: State): Promise<State> => {
    console.log('Thinking about next action...');
    state.messages.push({
      role: 'assistant',
      content: 'Let me analyze the current situation and decide on next steps.',
    });
    state.data.thoughtCount = (state.data.thoughtCount || 0) + 1;
    return state;
  };

  const act = async (state: State): Promise<State> => {
    console.log('Executing action...');
    state.messages.push({
      role: 'action',
      content: 'Executing: search for relevant information',
    });
    state.data.actionTaken = true;
    return state;
  };

  const observe = async (state: State): Promise<State> => {
    console.log('Observing results...');
    state.messages.push({
      role: 'observation',
      content: 'Found relevant information successfully.',
    });
    state.data.observationCount = (state.data.observationCount || 0) + 1;
    return state;
  };

  const respond = async (state: State): Promise<State> => {
    console.log('Generating final response...');
    state.messages.push({
      role: 'assistant',
      content: 'Based on my research, here is the answer...',
    });
    state.data.complete = true;
    return state;
  };

  // Build the graph
  const graph = new AegisGraph(run)
    .addNode('start', think)
    .addNode('act', act)
    .addNode('observe', observe)
    .addNode('respond', respond)
    .addEdge('start', 'act')
    .addEdge('act', 'observe')
    .addConditionalEdge('observe', (state) => {
      // If we've done enough iterations, respond; otherwise think again
      if ((state.data.observationCount || 0) >= 2) {
        return 'respond';
      }
      return 'start'; // Loop back for more thinking
    })
    .addEdge('respond', 'END');

  // Execute the graph
  const compiled = graph.compile();
  const finalState = await compiled.invoke({
    messages: [],
    currentStep: 'start',
    data: {},
  });

  console.log('\n=== Final State ===');
  console.log('Messages:', finalState.messages.length);
  console.log('Thought iterations:', finalState.data.thoughtCount);
  console.log('Observations:', finalState.data.observationCount);
  console.log('Complete:', finalState.data.complete);

  run.end({ state: finalState });
  console.log(`\nRun completed: ${run.runId}`);
}

// Example: Plan-Execute-Reflect pattern
async function examplePlanExecuteReflect() {
  const client = new AegisRunClient({ baseUrl: 'http://localhost:8080' });

  const run = await new Run({
    client,
    agentId: 'plan-execute-agent',
    policyRef: 'plan-execute-policy:v1',
  }).start();

  const plan = async (state: State): Promise<State> => {
    console.log('Planning...');
    state.messages.push({ role: 'plan', content: 'Created 3-step plan' });
    state.data.plan = ['step1', 'step2', 'step3'];
    return state;
  };

  const execute = async (state: State): Promise<State> => {
    console.log('Executing plan...');
    const currentPlanStep = state.data.currentPlanIndex || 0;
    state.messages.push({
      role: 'execute',
      content: `Executing: ${state.data.plan[currentPlanStep]}`,
    });
    state.data.currentPlanIndex = currentPlanStep + 1;
    return state;
  };

  const reflect = async (state: State): Promise<State> => {
    console.log('Reflecting on execution...');
    state.messages.push({
      role: 'reflect',
      content: 'Execution successful, proceeding...',
    });
    return state;
  };

  const graph = new AegisGraph(run)
    .addNode('start', plan)
    .addNode('execute', execute)
    .addNode('reflect', reflect)
    .addEdge('start', 'execute')
    .addEdge('execute', 'reflect')
    .addConditionalEdge('reflect', (state) => {
      // Continue executing until plan is complete
      if (state.data.currentPlanIndex < state.data.plan.length) {
        return 'execute';
      }
      return 'END';
    });

  const compiled = graph.compile();
  const finalState = await compiled.invoke({
    messages: [],
    currentStep: 'start',
    data: {},
  });

  console.log('\nFinal plan execution:', finalState);
  run.end({ outcome: finalState });
}

// Export for use as a module
export { AegisGraph, CompiledGraph, State, NodeFunction };

// Run example if executed directly
if (require.main === module) {
  exampleReActAgent().catch(console.error);
}

File: sdk/python/examples/langgraph_wrapper.py
"""LangGraph-style wrapper for AegisRun"""

from typing import Callable, Dict, Any, TypedDict
from aegisrun import Run, Step

class State(TypedDict):
    """Agent state"""

    messages: list
    current_step: str
    data: Dict[str, Any]

class AegisGraph:
    """LangGraph-like wrapper with AegisRun enforcement"""

    def __init__(self, run: Run):
        self.run = run
        self.nodes: Dict[str, Callable] = {}
        self.edges: Dict[str, str] = {}

    def add_node(self, name: str, fn: Callable[[State], State]):
        """Add a node to the graph"""
        self.nodes[name] = fn
        return self

    def add_edge(self, from_node: str, to_node: str):
        """Add an edge between nodes"""
        self.edges[from_node] = to_node
        return self

    def compile(self) -> "CompiledGraph":
        """Compile the graph"""
        return CompiledGraph(self.run, self.nodes, self.edges)

class CompiledGraph:
    """Compiled graph ready for execution"""

    def __init__(
        self, run: Run, nodes: Dict[str, Callable], edges: Dict[str, str]
    ):
        self.run = run
        self.nodes = nodes
        self.edges = edges

    def invoke(self, initial_state: State) -> State:
        """Execute the graph"""
        state = initial_state
        current_node = "start"

        while current_node in self.nodes:
            node_fn = self.nodes[current_node]

            # Execute node as a step
            def execute_node(step: Step):
                return node_fn(state)

            state = self.run.step(
                name=current_node,
                state_vector={"node": current_node, "state": state},
                fn=execute_node,
            )

            # Get next node
            current_node = self.edges.get(current_node)
            if not current_node:
                break

        return state

def example_usage():
    """Example LangGraph-style usage"""
    from aegisrun import AegisRunClient

    client = AegisRunClient(base_url="http://localhost:8080")

    run = Run(
        client=client,
        agent_id="langgraph-agent",
        policy_ref="langgraph-demo",
    ).start()

    # Define nodes
    def plan(state: State) -> State:
        print("Planning...")
        state["messages"].append({"role": "plan", "content": "Create a plan"})
        return state

    def execute(state: State) -> State:
        print("Executing...")
        state["messages"].append({"role": "execute", "content": "Execute plan"})
        return state

    def reflect(state: State) -> State:
        print("Reflecting...")
        state["messages"].append({"role": "reflect", "content": "Reflect on results"})
        return state

    # Build graph
    graph = (
        AegisGraph(run)
        .add_node("start", plan)
        .add_node("execute", execute)
        .add_node("reflect", reflect)
        .add_edge("start", "execute")
        .add_edge("execute", "reflect")
    )

    # Compile and run
    compiled = graph.compile()
    final_state = compiled.invoke(
        State(messages=[], current_step="start", data={})
    )

    print(f"Final state: {final_state}")

    run.end(outcome={"state": final_state})

if __name__ == "__main__":
    example_usage()
